language/evm/move-to-yul/src/events.rs

```diff
@@ -186,6 +186,163 @@ impl EventSignature {
     }
 }
 
+/// Generate emit functions for send_<name>
+pub(crate) fn define_emit_fun_for_send(
+    gen: &mut FunctionGenerator,
+    ctx: &Context,
+    event_sig: &EventSignature,
+    fun_id: &QualifiedInstId<FunId>,
+) {
+    let fun = ctx.env.get_function(fun_id.to_qualified_id());
+    let target = &ctx.targets.get_target(&fun, &FunctionVariant::Baseline);
+
+    // Emit function header
+    let params = (0..target.get_parameter_count()).map(|idx| ctx.make_local_name(target, idx));
+    let params_str = params.clone().join(",");
+    let params_vec = params.collect_vec();
+
+    let mut local_name_idx = target.get_parameter_count();
+
+    emit!(
+        ctx.writer,
+        "function {}({}) ",
+        ctx.make_function_name(fun_id),
+        params_str
+    );
+
+    // Generate the function body
+    ctx.emit_block(|| {
+        let signature_types = &event_sig.para_types;
+        let topic_0_var = ctx.make_local_name(target, local_name_idx);
+        local_name_idx += 1;
+        let event_sig_str = event_sig.to_string();
+        let topic_0_hash = format!("0x{:x}", Keccak256::digest(event_sig_str.as_bytes()));
+
+        emitln!(ctx.writer, "let {} := {}", topic_0_var, topic_0_hash);
+        let message_str = format!(
+            "{}::{}",
+            fun.module_env.self_address(),
+            fun.get_full_name_str().replace("send_", "")
+        );
+
+        let hash_bytes = Sha3_256::digest(message_str.as_bytes());
```

Comment:
Do you mind explain a bit how this part of code works? Specifically how will the event be encodeded? Maybe some comments here would be great.